### PR TITLE
[metricbeat] replace depricated io/ioutil

### PR DESCRIPTION
**Pull request by florianl**
## Proposed commit message

As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred. Successor functionality in io or os is more performant than the deprecated functionality in io/ioutil.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [ ] ~~I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

File: metricbeat/module/openmetrics/collector/_meta/data.json [hunk 1]
```diff
@@ -10,15 +10,18 @@
...
+        "help": "Total number of connections closed that were made to the listener of a given name.",
         "labels": {
-            "job": "openmetrics"
+            "job": "openmetrics",
+            "listener_name": "http"
         },
         "metrics": {
-            "up": 1
-        }
+            "net_conntrack_listener_conn_closed_total": 0
```
Comment (faec) [inline metricbeat/module/openmetrics/collector/_meta/data.json:19]: Was this included by accident? Everything else in the PR looks like natural in-place substitutions, but this seems unrelated. (If the `data.json` changes weren't here I would approve it.)

Comment (florianl) [reply metricbeat/module/openmetrics/collector/_meta/data.json:19]: sorry - I must have added them by accident and will remove them.

Comment (florianl) [reply metricbeat/module/openmetrics/collector/_meta/data.json]: file got removed from the PR.

**Review by faec — APPROVED** [review]

File: metricbeat/helper/server/http/http_test.go [hunk 2]
```diff
@@ -212,6 +212,10 @@ func writeToServer(t *testing.T, message, host string, port int, connectionMetho
 	url := fmt.Sprintf("%s://%s:%d/", strings.ToLower(connectionType), host, port)
 	var str = []byte(message)
 	req, err := http.NewRequest(connectionMethod, url, bytes.NewBuffer(str))
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
```
Comment (belimawr) [inline metricbeat/helper/server/http/http_test.go:217]: That can be replaced by a single `t.Fatal()`
```suggestion
		t.Fatal(err)
```

File: metricbeat/mb/testing/testdata.go [hunk 3]
```diff
@@ -362,7 +362,11 @@ func writeDataJSON(t *testing.T, data mapstr.M, path string) {
 	// Add hardcoded timestamp
 	data.Put("@timestamp", "2019-03-01T08:05:34.853Z")
 	output, err := json.MarshalIndent(&data, "", "    ")
-	if err = ioutil.WriteFile(path, output, 0644); err != nil {
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
```
Comment (belimawr) [inline metricbeat/mb/testing/testdata.go:367]: Same here
```suggestion
		t.Fatal(err)
```

File: metricbeat/module/ceph/monitor_health/monitor_health_test.go [hunk 4]
```diff
@@ -36,7 +36,8 @@ func TestFetchEventContents(t *testing.T) {
 	absPath, err := filepath.Abs("../_meta/testdata/")
 	assert.NoError(t, err)
 
-	response, err := ioutil.ReadFile(absPath + "/sample_response.json")
+	response, err := os.ReadFile(absPath + "/sample_response.json")
+	assert.NoError(t, err)
```
Comment (belimawr) [inline metricbeat/module/ceph/monitor_health/monitor_health_test.go:40]: It's better to use `require` here as it will immediately fail the test. If we fail to read the file, there  is no point continuing with the test.
```suggestion
	require.NoError(t, err)
```

**Review by belimawr — CHANGES_REQUESTED** [review]
Great job! I'm always happy to see this kin of improvements :heart: 

The only blocking suggestion is the assert/require one.

File: metricbeat/module/ceph/monitor_health/monitor_health_test.go (hunk 4 above)
Comment (florianl) [reply metricbeat/module/ceph/monitor_health/monitor_health_test.go:40]: What is the difference to line 37 `assert.NoError(t, err)` ? Why should we use here `require` instead of `assert`?

File: metricbeat/mb/testing/testdata.go (hunk 3 above)
Comment (florianl) [reply metricbeat/mb/testing/testdata.go:367]: Looking at existing code, `t.Error(err)` followed by `t.FailNow()` is the most common pattern. So far, no one cared that the error was ignored so far. And the change was added, just to make linter happy.

File: metricbeat/module/ceph/monitor_health/monitor_health_test.go (hunk 4 above)
Comment (belimawr) [reply metricbeat/module/ceph/monitor_health/monitor_health_test.go:40]: `require` calls `t.Fatal` or `t.FailNow` which effectively interrupts the test right when the assertion fails. In that case if we continue the test after `os.ReadFile` fails, the test will only produce more errors that are not helpful, hence stopping the test when the file cannot be read is the best approach in my opinion.

That is the only comment/change I consider a blocker, and even that is not a hard blocker.

File: metricbeat/mb/testing/testdata.go (hunk 3 above)
Comment (belimawr) [reply metricbeat/mb/testing/testdata.go:367]: > Looking at existing code, t.Error(err) followed by t.FailNow() is the most common pattern.

The bits of the codebase I mostly work with follow a different pattern 😅 , anyway this one is not a blocker.

File: metricbeat/module/ceph/monitor_health/monitor_health_test.go (hunk 4 above)
Comment (florianl) [reply metricbeat/module/ceph/monitor_health/monitor_health_test.go:40]: > That is the only comment/change I consider a blocker, and even that is not a hard blocker.

I can't follow why this is considered a blocker. Before this change `err` was never checked before and the discussion `assert` vs `require` is a style discussion, that should not affect this PR.

**Review by belimawr — APPROVED** [review]

**Review by klacabane — APPROVED** [review]

File: metricbeat/module/nats/connection/connection_test.go [hunk 5]
```diff
@@ -21,6 +21,7 @@ import (
 	"io/ioutil"
```
Comment (gizas) [inline metricbeat/module/nats/connection/connection_test.go:21]: So this can be removed?

**Review by gizas — APPROVED** [review]

File: metricbeat/module/nats/connection/connection_test.go (hunk 5 above)
Comment (florianl) [reply metricbeat/module/nats/connection/connection_test.go]: good catch :+1: sorry that I have missed this! It got removed with https://github.com/elastic/beats/pull/36963/commits/937a28b4c1df6af34f7e5d6d60fee3f5d56002b6

**Comment by florianl** [comment]
satisfying linter is a major hurdle to get this merged. as I don't have capacity to do so, I'm closing this PR.
